feat: add docs on remote deployments multiple device services in secure (#1278)
edgexfoundry/edgex-docs: docs_src/microservices/device/remote/Ch-RemoteSecure.md

    !!! note
        Refer to sample [device virtual service deployment in secure mode](../../../security/Ch-RemoteDeviceServices.md) for general understanding

1. Set up the "local" node to be ready for deployment. Refer to [USB Service Setup](../services/device-usb-camera/walkthrough/setup.md)
   for system requirements and dependencies such as Git, Docker, Docker Compose, etc.

2. Next step is to install [EdgeX compose](https://github.com/edgexfoundry/edgex-compose) in the host node which will be used to run all EdgeX core services. So clone the `edgex-compose`
   repository:

     ```bash
     git clone https://github.com/edgexfoundry/edgex-compose.git
     ```

3. Checkout the required version:

      ```bash
      git checkout {{edgexversion}}
      ```

### Generate SSH key pair for SSH tunnel
1. Navigate to `edgex-examples/security/remote_devices/spiffe_and_ssh`
2. Run `generate_keys.sh` script
3. Copy `remote` folder located at `edgex-examples/security/remote_devices/spiffe_and_ssh` to remote nodes

### Edit docker-compose file
1. Navigate to `edgex-examples/security/remote_devices/spiffe_and_ssh/local` and edit
   `docker-compose.yml` file
 
    !!! note

>         Below instructions explains to run 2 instances of usb service: instance 1 and instance 2)

Reviewer: ```suggestion
        Below instructions explains to run 2 instances of usb service: instance 1 and instance 2
```

2. Change `host_ip` to ip address of local system for `consul`, `core-metadata`, `database` and `vault`

```yaml
services:
  consul:
    ports:
      - mode: ingress
        host_ip: 192.168.4.103    # insert local machine's host ip here
        target: 8500
        published: "8500"
        protocol: tcp

  core-metadata:
    ports:
      - mode: ingress
        host_ip: 192.168.4.103    # insert local machine's host ip here
        target: 59881
        published: "59881"
        protocol: tcp
  database:
    ports:
      - mode: ingress
        host_ip: 192.168.4.103    # insert local machine's host ip here
        target: 6379
        published: "6379"
        protocol: tcp
  vault:
    ports:
      - mode: ingress
        host_ip: 192.168.4.103    # insert local machine's host ip here
        target: 8200
        published: "8200"
        protocol: tcp
```

> 1. Add/edit environment variables for `consul`, `security-proxy-setup` and `security-secretstore-setup` services

Reviewer: There is some issue with numbering here. Render and test it.

```yaml
  services:
    consul:
      environment:
        EDGEX_ADD_REGISTRY_ACL_ROLES: "device-usb-camera_1,device-usb-camera_2"

    security-proxy-setup:
      environment:
        EDGEX_ADD_PROXY_ROUTE: "device-usb-camera_1.http://192.168.4.104:59983,device-usb-camera_2.http://192.168.4.105:59983"
        ROUTES_DEVICE_USB_CAMERAL_HOST: "device-usb-camera_1,device-usb-camera_2"

    security-secretstore-setup:
      environment:
        EDGEX_ADD_KNOWN_SECRETS: "redisdb[app-rules-engine],redisdb[device-usb-camera_1],message-bus[device-usb-camera_1],redisdb[device-usb-camera_2],message-bus[device-usb-camera_2]"
```

> 1. Add multiple `device-ssh-proxy` service and configure with remote node ips

Reviewer: Also issue with numbering here and down below.

>    - copy `device-ssh-proxy` service and paste it again below itself.

Reviewer: ```suggestion
   - Copy `device-ssh-proxy` service and paste it again below itself
```

>    - rename services to `device-ssh-proxy-1` and `device-ssh-proxy-2` respectively

Reviewer: ```suggestion
   - Rename services to `device-ssh-proxy-1` and `device-ssh-proxy-2` respectively
```

   - Modify the existing environment variables and other fields for `device-ssh-proxy-1`
   
   ```yaml
     services:
       device-ssh-proxy-1:
         container_name: edgex-device-ssh-proxy-1
         environment:
           SERVICE_HOST: edgex-device-usb-camera

>            SERVICE_PORT: 59983

Author: I do not believe so, as the steps show port mapping of 59984 external to 59983 internal. I'm not sure if it would make things simpler if the internal and external were both changed.

>            SERVICE_PORT: 59983

Reviewer 2: Right that is changing the internal.

         hostname: edgex-device-ssh-proxy-2
         networks:
           edgex-network:
             aliases:
               - edgex-device-usb-camera-2
         ports:
           - 127.0.0.1:59984:59983/tcp            # notice the different port mapping to avoid conflicts
   ```

2. Update to latest docker image for `security-spiffe-token-provider`

```yaml
  services:
    security-spiffe-token-provider:
      image: nexus3.edgexfoundry.org:10004/security-spiffe-token-provider:latest
```

### Build and run services
1. Navigate to edgex-examples/security/remote_devices/spiffe_and_ssh/lcoal
2. Build and start the services
    ```bash
    docker compose build
    docker compose up -d
    ```
3. Run `docker ps -a` and ensure all containers started without crashing (Note:
   `edgex-device-ssh-proxy-1` and `edgex-device-ssh-proxy-2` will keep restarting until services on remote node start.)

### Update server entries
Edit and run `add-server-entry.sh` script for each usb service

1. Navigate to `edgex-examples/security/remote_devices/spiffe_and_ssh/lcoal`
   and edit `add-server-entry.sh` script like below.

>    !!! example - Example for first device service

Reviewer: This snippet does not render properly.

       ```bash
       docker exec -ti edgex-security-spire-config spire-server entry create \
           -socketPath /tmp/edgex/secrets/spiffe/private/api.sock  \
           -parentID spiffe://edgexfoundry.org/spire/agent/x509pop/cn/  \
           remote-agent -dns "edgex-device-usb-camera" \
           -spiffeID  spiffe://edgexfoundry.org/service/device-usb-camera_1 \
           -selector "docker:label:com.docker.compose.service:device-usb-camera_1"
       ```

2. Save and then run `./add-server-entry.sh`

3. Repeat previous steps for each additional device service

>    !!! example - Example for second device service

Reviewer: This snippet does not render properly.

       ```bash
       docker exec -ti edgex-security-spire-config spire-server entry create \
           -socketPath /tmp/edgex/secrets/spiffe/private/api.sock  \
           -parentID spiffe://edgexfoundry.org/spire/agent/x509pop/cn/  \
           remote-agent -dns "edgex-device-usb-camera" \
           -spiffeID  spiffe://edgexfoundry.org/service/device-usb-camera_2 \
           -selector "docker:label:com.docker.compose.service:device-usb-camera_2"
       ```
4. Save and then run `./add-server-entry.sh` again

> ## Remote Node Setup

Reviewer 2: Clarify that this needs to be done once on each remote node.

1. Navigate to `edgex-examples/security/remote_devices/spiffe_and_ssh/remote`
   and update `docker-compose.yml`. 

> 2. Remove `device-virtual` service and add 2 `device-usb-service` like below, one for each instance.

Reviewer 2: ```suggestion
2. Remove `device-virtual` service and add `device-usb-service` like below, one for each instance on each node.
```

             SECRETSTORE_RUNTIMETOKENPROVIDER_REQUIREDSECRETS: redisdb
             SECRETSTORE_RUNTIMETOKENPROVIDER_TRUSTDOMAIN: edgexfoundry.org
             DEVICE_DISCOVERY_ENABLED: 'true' # enable or disable auto-discovery
             DEVICE_DISCOVERY_INTERVAL: '5m'  # configure the auto-discovery interval
           hostname: edgex-device-usb-camera
           image: nexus3.edgexfoundry.org:10004/device-usb-camera:latest
           networks:
            edgex-network: {}
           ports:
             - "192.168.4.105:59983:59983/tcp"   # <remote node ip>:59983:59983/tcp
             - "8554:8554"
             - "8000:8000"
           read_only: true
           restart: always
           security_opt:
             - no-new-privileges:true
           user: 'root:root'
           volumes:
             # - /tmp/edgex/secrets/device-usb-camera_<instance number>:/tmp/edgex/secrets/device-usb-camera_<instance number>:ro,z
             - /tmp/edgex/secrets/device-usb-camera_1:/tmp/edgex/secrets/device-usb-camera_1:ro,z
             - /tmp/edgex/secrets/spiffe/public:/tmp/edgex/secrets/spiffe/public:ro,z
             - /dev:/dev:ro
             - /run/udev:/run/udev:ro
    ```

> 3. Build and run service on remote node

Reviewer 2: ```suggestion
3. Build and run services on each remote node
```

   - Navigate to `edgex-examples/security/remote_devices/spiffe_and_ssh/remote`
   - `docker compose build`
   - `docker compose up -d`
   - Run `docker ps- a` to ensure all services started successfully
4) Wait for about 1 minute to establish ssh tunnel and to establish communication
   between edgex server system and remote node
5. Run `docker logs edgex-device-usb-camera` to check usb service status. Ensure
   usb service started successfully and detected and added connected usb cameras (if
   any) to `edgex-core-metadata`

> Note: If edgex services are restarted then remote node services must also be

Reviewer: Replace this note to be like others to have consistency.

restarted

## Testing and execution of REST APIs
1. Download [get-api-gateway-token.sh](https://github.com/edgexfoundry/edgex-compose/blob/main/compose-builder/get-api-gateway-token.sh) script

> 2. Navigate to `edgex-examples/security/remote_devices/spiffe_and_ssh/lcoal` and

Reviewer: ```suggestion
2. Navigate to `edgex-examples/security/remote_devices/spiffe_and_ssh/local` and
```
